iothub_client/samples/iothub_client_device_twin_and_methods_sample/iothub_client_device_twin_and_methods_sample.c

```diff
@@ -32,170 +42,550 @@
     {
-        (void)memset(car, 0, sizeof(Car));
+        (void)printf("Failed to send a GET request for the twin document asynchronously: return code %d.\n", rc);
+        exit(rc);
+    }
 
-        root_value = json_parse_string(json);
-        root_object = json_value_get_object(root_value);
+    // Set callback for twin desired property PATCH messages from IoT Hub.
+    // Sends GET request for twin document as part of setting the callback.
+    rc = IoTHubDeviceClient_SetDeviceTwinCallback(iothub_client, patch_twin_desired_properties_callback, NULL);
+    if (rc != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("Failed to set the twin document callback: return code %d.\n", rc);
+        exit(rc);
+    }
 
-        // Only desired properties:
-        JSON_Value* changeOilReminder;
-        JSON_Value* desired_maxSpeed;
-        JSON_Value* latitude;
-        JSON_Value* longitude;
+    // Set callback for C2D direct methods from the service client application.
+    rc = IoTHubDeviceClient_SetDeviceMethodCallback(iothub_client, receive_method_callback, NULL);
+    if (rc != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("Failed to set the C2D method callback: return code %d.\n", rc);
+        exit(rc);
+    }
 
-        if (update_state == DEVICE_TWIN_UPDATE_COMPLETE)
-        {
-            changeOilReminder = json_object_dotget_value(root_object, "desired.changeOilReminder");
-            desired_maxSpeed = json_object_dotget_value(root_object, "desired.settings.desired_maxSpeed");
-            latitude = json_object_dotget_value(root_object, "desired.settings.location.latitude");
-            longitude = json_object_dotget_value(root_object, "desired.settings.location.longitude");
-        }
-        else
-        {
-            changeOilReminder = json_object_dotget_value(root_object, "changeOilReminder");
-            desired_maxSpeed = json_object_dotget_value(root_object, "settings.desired_maxSpeed");
-            latitude = json_object_dotget_value(root_object, "settings.location.latitude");
-            longitude = json_object_dotget_value(root_object, "settings.location.longitude");
-        }
+    // Wait for twin` GET response, a reported property PATCH response, a desired property PATCH
+    // message, or a direct method. Returns if timeout occurs. After TIMEOUT_SEND_GET_REQUEST_MS
+    // seconds, request the twin document's desired properties.
+    time_t receive_message_start_time = time(NULL);
+#if defined SAMPLE_MQTT || defined SAMPLE_MQTT_OVER_WEBSOCKETS
+    time_t send_get_request_start_time = time(NULL);
+#endif
+    time_t current_time;
+    message_received = false;
 
-        if (changeOilReminder != NULL)
+    for (uint8_t message_count = 0; message_count < MAX_MESSAGE_COUNT; message_count++)
+    {
+        // Wait up to TIMEOUT_RECEIVE_MS to receive a message.
+        while(1)
         {
-            const char* data = json_value_get_string(changeOilReminder);
+            if (message_received)
```

Comment:
can't u refactor with message_received in the while check and move receive_message_start_time = time(NULL); and message_received = false; outside the while()?

Reply:
add a tick delay to the loop?